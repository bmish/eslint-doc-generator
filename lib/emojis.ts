import { SEVERITY_TYPE } from './types.js';
import { EMOJIS_TYPE } from './rule-type.js';

// Configs.
const EMOJI_A11Y = '‚ôø';
const EMOJI_ERROR = '‚ùó';
const EMOJI_STYLE = 'üé®';
const EMOJI_TYPESCRIPT = '‚å®Ô∏è';
const EMOJI_WARNING = 'üö∏';
/** Default emojis for common configs. */
export const EMOJI_CONFIGS = {
  a11y: EMOJI_A11Y,
  accessibility: EMOJI_A11Y,
  all: 'üåê',
  error: EMOJI_ERROR,
  errors: EMOJI_ERROR,
  recommended: '‚úÖ',
  'recommended-type-aware': '‚òëÔ∏è',
  'recommended-type-checked': '‚òëÔ∏è',
  strict: 'üîí',
  style: EMOJI_STYLE,
  stylistic: EMOJI_STYLE,
  ts: EMOJI_TYPESCRIPT,
  type: EMOJI_TYPESCRIPT,
  'type-aware': EMOJI_TYPESCRIPT,
  'type-checked': EMOJI_TYPESCRIPT,
  typed: EMOJI_TYPESCRIPT,
  types: EMOJI_TYPESCRIPT,
  typescript: EMOJI_TYPESCRIPT,
  warning: EMOJI_WARNING,
  warnings: EMOJI_WARNING,
};

// Severities.
export const EMOJI_CONFIG_ERROR = 'üíº';
export const EMOJI_CONFIG_WARN = '‚ö†Ô∏è';
export const EMOJI_CONFIG_OFF = 'üö´';
/** Emoji for each config severity. */
export const EMOJI_CONFIG_FROM_SEVERITY: {
  [key in SEVERITY_TYPE]: string;
} = {
  [SEVERITY_TYPE.error]: EMOJI_CONFIG_ERROR,
  [SEVERITY_TYPE.warn]: EMOJI_CONFIG_WARN,
  [SEVERITY_TYPE.off]: EMOJI_CONFIG_OFF,
};

/** Rule has an autofixer (from `meta.fixable`). */
export const EMOJI_FIXABLE = 'üîß';

/** Rule provides suggestions (`meta.hasSuggestions`). */
export const EMOJI_HAS_SUGGESTIONS = 'üí°';

/** Rule options (from `meta.schema`). */
export const EMOJI_OPTIONS = '‚öôÔ∏è';

/**
 * Rule requires type-checking (from `meta.docs.requiresTypeChecking`).
 * Should match the emoji that @typescript-eslint/eslint-plugin uses for this (https://typescript-eslint.io/rules/).
 */
export const EMOJI_REQUIRES_TYPE_CHECKING = 'üí≠';

/**
 * Rule type (from `meta.type`).
 * Also see EMOJIS_TYPE defined in rule-type.ts.
 */
export const EMOJI_TYPE = 'üóÇÔ∏è';

/** Rule is deprecated (from `meta.deprecated`). */
export const EMOJI_DEPRECATED = '‚ùå';

/**
 * The user is not allowed to specify a reserved emoji to represent their config because we use these emojis for other purposes.
 * Note that the default emojis for common configs are intentionally not reserved.
 */
export const RESERVED_EMOJIS = [
  ...Object.values(EMOJI_CONFIG_FROM_SEVERITY),
  ...Object.values(EMOJIS_TYPE),

  EMOJI_FIXABLE,
  EMOJI_HAS_SUGGESTIONS,
  EMOJI_OPTIONS,
  EMOJI_REQUIRES_TYPE_CHECKING,
  EMOJI_TYPE,
  EMOJI_DEPRECATED,
];
