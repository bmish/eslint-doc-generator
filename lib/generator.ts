import { EOL } from 'node:os';
import { existsSync, mkdirSync, readFileSync, writeFileSync } from 'node:fs';
import { dirname, join, relative, resolve } from 'node:path';
import { getAllNamedOptions, hasOptions } from './rule-options.js';
import {
  loadPlugin,
  getPluginPrefix,
  getPluginRoot,
  getPathWithExactFileNameCasing,
} from './package-json.js';
import { updateRulesList } from './rule-list.js';
import { updateConfigsList } from './config-list.js';
import { generateRuleHeaderLines } from './rule-doc-notices.js';
import {
  parseRuleDocNoticesOption,
  parseRuleListColumnsOption,
  parseConfigEmojiOptions,
} from './option-parsers.js';
import {
  BEGIN_RULE_OPTIONS_LIST_MARKER,
  END_RULE_HEADER_MARKER,
  END_RULE_OPTIONS_LIST_MARKER,
} from './comment-markers.js';
import {
  replaceOrCreateHeader,
  expectContentOrFail,
  expectSectionHeaderOrFail,
} from './markdown.js';
import { resolveConfigsToRules } from './plugin-config-resolution.js';
import { OPTION_DEFAULTS } from './options.js';
import { diff } from 'jest-diff';
import type { GenerateOptions } from './types.js';
import { OPTION_TYPE, RuleModule } from './types.js';
import { replaceRulePlaceholder } from './rule-link.js';
import { updateRuleOptionsList } from './rule-options-list.js';

function stringOrArrayWithFallback<T extends string | readonly string[]>(
  stringOrArray: undefined | T,
  fallback: T
): T {
  return stringOrArray && stringOrArray.length > 0 ? stringOrArray : fallback;
}

function stringOrArrayToArrayWithFallback(
  stringOrArray: undefined | string | readonly string[],
  fallback: readonly string[]
): readonly string[] {
  const asArray =
    stringOrArray instanceof Array // eslint-disable-line unicorn/no-instanceof-array -- using Array.isArray() loses type information about the array.
      ? stringOrArray
      : stringOrArray
      ? [stringOrArray]
      : [];
  const csvStringItem = asArray.find((item) => item.includes(','));
  if (csvStringItem) {
    throw new Error(
      `Provide property as array, not a CSV string: ${csvStringItem}`
    );
  }
  return asArray && asArray.length > 0 ? asArray : fallback;
}

// eslint-disable-next-line complexity
export async function generate(path: string, options?: GenerateOptions) {
  const plugin = await loadPlugin(path);
  const pluginPrefix = getPluginPrefix(path);
  const configsToRules = await resolveConfigsToRules(plugin);

  if (!plugin.rules) {
    throw new Error('Could not find exported `rules` object in ESLint plugin.');
  }

  // Options. Add default values as needed.
  const check = options?.check ?? OPTION_DEFAULTS[OPTION_TYPE.CHECK];
  const configEmojis = parseConfigEmojiOptions(plugin, options?.configEmoji);
  const configFormat =
    options?.configFormat ?? OPTION_DEFAULTS[OPTION_TYPE.CONFIG_FORMAT];
  const ignoreConfig = stringOrArrayWithFallback(
    options?.ignoreConfig,
    OPTION_DEFAULTS[OPTION_TYPE.IGNORE_CONFIG]
  );
  const ignoreDeprecatedRules =
    options?.ignoreDeprecatedRules ??
    OPTION_DEFAULTS[OPTION_TYPE.IGNORE_DEPRECATED_RULES];
  const initRuleDocs =
    options?.initRuleDocs ?? OPTION_DEFAULTS[OPTION_TYPE.INIT_RULE_DOCS];
  const pathRuleDoc =
    options?.pathRuleDoc ?? OPTION_DEFAULTS[OPTION_TYPE.PATH_RULE_DOC];
  const pathRuleList = stringOrArrayToArrayWithFallback(
    options?.pathRuleList,
    OPTION_DEFAULTS[OPTION_TYPE.PATH_RULE_LIST]
  );
  const postprocess =
    options?.postprocess ?? OPTION_DEFAULTS[OPTION_TYPE.POSTPROCESS];
  const ruleDocNotices = parseRuleDocNoticesOption(options?.ruleDocNotices);
  const ruleDocSectionExclude = stringOrArrayWithFallback(
    options?.ruleDocSectionExclude,
    OPTION_DEFAULTS[OPTION_TYPE.RULE_DOC_SECTION_EXCLUDE]
  );
  const ruleDocSectionInclude = stringOrArrayWithFallback(
    options?.ruleDocSectionInclude,
    OPTION_DEFAULTS[OPTION_TYPE.RULE_DOC_SECTION_INCLUDE]
  );
  const ruleDocSectionOptions =
    options?.ruleDocSectionOptions ??
    OPTION_DEFAULTS[OPTION_TYPE.RULE_DOC_SECTION_OPTIONS];
  const ruleDocTitleFormat =
    options?.ruleDocTitleFormat ??
    OPTION_DEFAULTS[OPTION_TYPE.RULE_DOC_TITLE_FORMAT];
  const ruleListColumns = parseRuleListColumnsOption(options?.ruleListColumns);
  const ruleListSplit =
    typeof options?.ruleListSplit === 'function'
      ? options.ruleListSplit
      : stringOrArrayToArrayWithFallback(
          options?.ruleListSplit,
          OPTION_DEFAULTS[OPTION_TYPE.RULE_LIST_SPLIT]
        );
  const urlConfigs =
    options?.urlConfigs ?? OPTION_DEFAULTS[OPTION_TYPE.URL_CONFIGS];
  const urlRuleDoc =
    options?.urlRuleDoc ?? OPTION_DEFAULTS[OPTION_TYPE.URL_RULE_DOC];

  // Gather normalized list of rules.
  const ruleNamesAndRules = Object.entries(plugin.rules)
    .map(([name, ruleModule]) => {
      // Convert deprecated function-style rules to object-style rules so that we don't have to handle function-style rules everywhere throughout the codebase.
      // @ts-expect-error -- this type unfortunately requires us to choose a `meta.type` even though the deprecated function-style rule won't have one.
      const ruleModuleAsObject: RuleModule =
        typeof ruleModule === 'function'
          ? {
              // Deprecated function-style rule don't support most of the properties that object-style rules support, so we'll just use the bare minimum.
              meta: {
                // @ts-expect-error -- type is missing for this property
                schema: ruleModule.schema, // eslint-disable-line @typescript-eslint/no-unsafe-assignment -- type is missing for this property
                // @ts-expect-error -- type is missing for this property
                deprecated: ruleModule.deprecated, // eslint-disable-line @typescript-eslint/no-unsafe-assignment -- type is missing for this property
              },
              create: ruleModule,
            }
          : ruleModule;
      const tuple: [string, RuleModule] = [name, ruleModuleAsObject];
      return tuple;
    })
    .filter(
      // Filter out deprecated rules from being checked, displayed, or updated if the option is set.
      ([, rule]) => !ignoreDeprecatedRules || !rule.meta.deprecated
    )
    .sort(([a], [b]) => a.toLowerCase().localeCompare(b.toLowerCase()));

  // Update rule doc for each rule.
  let initializedRuleDoc = false;
  for (const [name, rule] of ruleNamesAndRules) {
    const schema = rule.meta?.schema;
    const description = rule.meta?.docs?.description;
    const pathToDoc = join(path, replaceRulePlaceholder(pathRuleDoc, name));
    const ruleHasOptions = hasOptions(schema);

    if (!existsSync(pathToDoc)) {
      if (!initRuleDocs) {
        throw new Error(
          `Could not find rule doc (run with --init-rule-docs to create): ${relative(
            getPluginRoot(path),
            pathToDoc
          )}`
        );
      }

      // Determine content for fresh rule doc, including any mandatory sections.
      // The rule doc header will be added later.
      let newRuleDocContents = [
        ruleDocSectionInclude.length > 0
          ? ruleDocSectionInclude
              .map((title) => `## ${title}`)
              .join(`${EOL}${EOL}`)
          : undefined,
        ruleHasOptions
          ? `## Options${EOL}${EOL}${BEGIN_RULE_OPTIONS_LIST_MARKER}${EOL}${END_RULE_OPTIONS_LIST_MARKER}`
          : undefined,
      ]
        .filter((section) => section !== undefined)
        .join(`${EOL}${EOL}`);
      if (newRuleDocContents !== '') {
        newRuleDocContents = `${EOL}${newRuleDocContents}${EOL}`;
      }

      mkdirSync(dirname(pathToDoc), { recursive: true });
      writeFileSync(pathToDoc, newRuleDocContents);
      initializedRuleDoc = true;
    }

    // Regenerate the header (title/notices) of each rule doc.
    const newHeaderLines = generateRuleHeaderLines(
      description,
      name,
      plugin,
      configsToRules,
      pluginPrefix,
      path,
      pathRuleDoc,
      configEmojis,
      configFormat,
      ignoreConfig,
      ruleDocNotices,
      ruleDocTitleFormat,
      urlConfigs,
      urlRuleDoc
    );

    const contentsOld = readFileSync(pathToDoc).toString();
    const contentsNew = await postprocess(
      updateRuleOptionsList(
        replaceOrCreateHeader(
          contentsOld,
          newHeaderLines,
          END_RULE_HEADER_MARKER
        ),
        rule
      ),
      resolve(pathToDoc)
    );

    if (check) {
      if (contentsNew !== contentsOld) {
        console.error(
          `Please run eslint-doc-generator. A rule doc is out-of-date: ${relative(
            getPluginRoot(path),
            pathToDoc
          )}`
        );
        console.error(diff(contentsNew, contentsOld, { expand: false }));
        process.exitCode = 1;
      }
    } else {
      writeFileSync(pathToDoc, contentsNew);
    }

    // Check for potential issues with the rule doc.

    // Check for required sections.
    for (const section of ruleDocSectionInclude) {
      expectSectionHeaderOrFail(
        `\`${name}\` rule doc`,
        contentsNew,
        [section],
        true
      );
    }

    // Check for disallowed sections.
    for (const section of ruleDocSectionExclude) {
      expectSectionHeaderOrFail(
        `\`${name}\` rule doc`,
        contentsNew,
        [section],
        false
      );
    }

    if (ruleDocSectionOptions) {
      // Options section.
      expectSectionHeaderOrFail(
        `\`${name}\` rule doc`,
        contentsNew,
        ['Options', 'Config'],
        ruleHasOptions
      );
      for (const { name: namedOption } of getAllNamedOptions(schema)) {
        expectContentOrFail(
          `\`${name}\` rule doc`,
          'rule option',
          contentsNew,
          namedOption,
          true
        ); // Each rule option is mentioned.
      }
    }
  }

  if (initRuleDocs && !initializedRuleDoc) {
    throw new Error(
      '--init-rule-docs was enabled, but no rule doc file needed to be created.'
    );
  }

  for (const pathRuleListItem of pathRuleList) {
    // Find the exact filename.
    const pathToFile = getPathWithExactFileNameCasing(
      join(path, pathRuleListItem)
    );
    if (!pathToFile || !existsSync(pathToFile)) {
      throw new Error(
        `Could not find ${String(pathRuleList)} in ESLint plugin.`
      );
    }

    // Update the rules list in this file.
    const fileContents = readFileSync(pathToFile, 'utf8');
    const fileContentsNew = await postprocess(
      updateConfigsList(
        updateRulesList(
          ruleNamesAndRules,
          fileContents,
          plugin,
          configsToRules,
          pluginPrefix,
          pathRuleDoc,
          pathToFile,
          path,
          configEmojis,
          configFormat,
          ignoreConfig,
          ruleListColumns,
          ruleListSplit,
          urlConfigs,
          urlRuleDoc
        ),
        plugin,
        configsToRules,
        pluginPrefix,
        configEmojis,
        configFormat,
        ignoreConfig
      ),
      resolve(pathToFile)
    );

    if (check) {
      if (fileContentsNew !== fileContents) {
        console.error(
          `Please run eslint-doc-generator. The rules table in ${relative(
            getPluginRoot(path),
            pathToFile
          )} is out-of-date.`
        );
        console.error(diff(fileContentsNew, fileContents, { expand: false }));
        process.exitCode = 1;
      }
    } else {
      writeFileSync(pathToFile, fileContentsNew, 'utf8');
    }
  }
}
