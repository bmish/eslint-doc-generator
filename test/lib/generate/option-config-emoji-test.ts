import { generate } from '../../../lib/generator.js';
import mockFs from 'mock-fs';
import { dirname, resolve } from 'node:path';
import { fileURLToPath } from 'node:url';
import { readFileSync } from 'node:fs';
import { jest } from '@jest/globals';
import { EMOJI_CONFIG_ERROR } from '../../../lib/emojis.js';

const __dirname = dirname(fileURLToPath(import.meta.url));

const PATH_NODE_MODULES = resolve(__dirname, '..', '..', '..', 'node_modules');

describe('generate (--config-emoji)', function () {
  describe('basic', function () {
    beforeEach(function () {
      mockFs({
        'package.json': JSON.stringify({
          name: 'eslint-plugin-test',
          exports: 'index.js',
          type: 'module',
        }),

        'index.js': `
              export default {
                rules: {
                  'no-foo': { meta: { docs: { description: 'Description for no-foo.'} }, create(context) {} },
                  'no-bar': { meta: { docs: { description: 'Description for no-bar.'} }, create(context) {} },
                  'no-baz': { meta: { docs: { description: 'Description for no-boz.'} }, create(context) {} },
                },
                configs: {
                  recommended: {
                    rules: { 'test/no-foo': 'error', 'test/no-bar': 'error' },
                  },
                  stylistic: {
                    rules: { 'test/no-bar': 'error', 'test/no-baz': 'error' },
                  },
                  configWithoutEmoji: {
                    rules: { 'test/no-baz': 'error' },
                  }
                }
              };`,

        'README.md': '## Rules\n',

        'docs/rules/no-foo.md': '',
        'docs/rules/no-bar.md': '',
        'docs/rules/no-baz.md': '',

        // Needed for some of the test infrastructure to work.
        node_modules: mockFs.load(PATH_NODE_MODULES),
      });
    });

    afterEach(function () {
      mockFs.restore();
      jest.resetModules();
    });

    it('shows the correct emojis', async function () {
      await generate('.', {
        configEmoji: [
          ['recommended', 'ðŸ”¥'],
          ['stylistic', 'ðŸŽ¨'],
        ],
      });
      expect(readFileSync('README.md', 'utf8')).toMatchSnapshot();
      expect(readFileSync('docs/rules/no-foo.md', 'utf8')).toMatchSnapshot();
      expect(readFileSync('docs/rules/no-bar.md', 'utf8')).toMatchSnapshot();
      expect(readFileSync('docs/rules/no-baz.md', 'utf8')).toMatchSnapshot();
    });
  });

  describe('invalid format', function () {
    beforeEach(function () {
      mockFs({
        'package.json': JSON.stringify({
          name: 'eslint-plugin-test',
          exports: 'index.js',
          type: 'module',
        }),

        'index.js': `
              export default {
                rules: {
                  'no-foo': { meta: { docs: { description: 'Description for no-foo.'} }, create(context) {} },
                },
              };`,

        'README.md': '## Rules\n',

        'docs/rules/no-foo.md': '',

        // Needed for some of the test infrastructure to work.
        node_modules: mockFs.load(PATH_NODE_MODULES),
      });
    });

    afterEach(function () {
      mockFs.restore();
      jest.resetModules();
    });

    it('throws an error', async function () {
      await expect(
        generate('.', { configEmoji: [['foo', 'bar', 'baz']] })
      ).rejects.toThrow(
        'Invalid configEmoji option: foo,bar,baz. Expected format: config,emoji'
      );
    });
  });

  describe('trying to remove a default emoji that does not exist', function () {
    beforeEach(function () {
      mockFs({
        'package.json': JSON.stringify({
          name: 'eslint-plugin-test',
          exports: 'index.js',
          type: 'module',
        }),

        'index.js': `
              export default {
                rules: {
                  'no-foo': { meta: { docs: { description: 'Description for no-foo.'} }, create(context) {} },
                },
              };`,

        'README.md': '## Rules\n',

        'docs/rules/no-foo.md': '',

        // Needed for some of the test infrastructure to work.
        node_modules: mockFs.load(PATH_NODE_MODULES),
      });
    });

    afterEach(function () {
      mockFs.restore();
      jest.resetModules();
    });

    it('throws an error', async function () {
      await expect(
        generate('.', { configEmoji: [['config-without-default-emoji']] })
      ).rejects.toThrow(
        'Invalid configEmoji option: config-without-default-emoji. Expected format: config,emoji'
      );
    });
  });

  describe('removing default emoji for a config', function () {
    beforeEach(function () {
      mockFs({
        'package.json': JSON.stringify({
          name: 'eslint-plugin-test',
          exports: 'index.js',
          type: 'module',
        }),

        'index.js': `
              export default {
                rules: {
                  'no-foo': { meta: { docs: { description: 'Description for no-foo.'} }, create(context) {} },
                },
                configs: {
                  recommended: { rules: { 'test/no-foo': 'error' } },
                }
              };`,

        'README.md': '## Rules\n',

        'docs/rules/no-foo.md': '',

        // Needed for some of the test infrastructure to work.
        node_modules: mockFs.load(PATH_NODE_MODULES),
      });
    });

    afterEach(function () {
      mockFs.restore();
      jest.resetModules();
    });

    it('reverts to using a badge for the config', async function () {
      await generate('.', {
        configEmoji: [['recommended']],
      });
      expect(readFileSync('README.md', 'utf8')).toMatchSnapshot();
      expect(readFileSync('docs/rules/no-foo.md', 'utf8')).toMatchSnapshot();
    });
  });

  describe('non-existent config', function () {
    beforeEach(function () {
      mockFs({
        'package.json': JSON.stringify({
          name: 'eslint-plugin-test',
          exports: 'index.js',
          type: 'module',
        }),

        'index.js': `
              export default {
                rules: {
                  'no-foo': { meta: { docs: { description: 'Description for no-foo.'} }, create(context) {} },
                },
                configs: {}
              };`,

        'README.md': '## Rules\n',

        'docs/rules/no-foo.md': '',

        // Needed for some of the test infrastructure to work.
        node_modules: mockFs.load(PATH_NODE_MODULES),
      });
    });

    afterEach(function () {
      mockFs.restore();
      jest.resetModules();
    });

    it('throws an error', async function () {
      await expect(
        generate('.', { configEmoji: [['config-does-not-exist', 'ðŸ”¥']] })
      ).rejects.toThrow(
        'Invalid configEmoji option: config-does-not-exist config not found.'
      );
    });
  });

  describe('using a reserved emoji', function () {
    beforeEach(function () {
      mockFs({
        'package.json': JSON.stringify({
          name: 'eslint-plugin-test',
          exports: 'index.js',
          type: 'module',
        }),

        'index.js': `
              export default {
                rules: {
                  'no-foo': { meta: { docs: { description: 'Description for no-foo.'} }, create(context) {} },
                },
                configs: {
                  recommended: { rules: { 'test/no-foo': 'error' } },
                  style: { rules: { 'test/no-foo': 'error' } },
                }
              };`,

        'README.md': '## Rules\n',

        'docs/rules/no-foo.md': '',

        // Needed for some of the test infrastructure to work.
        node_modules: mockFs.load(PATH_NODE_MODULES),
      });
    });

    afterEach(function () {
      mockFs.restore();
      jest.resetModules();
    });

    it('throws an error', async function () {
      await expect(
        generate('.', { configEmoji: [['recommended', EMOJI_CONFIG_ERROR]] })
      ).rejects.toThrow(`Cannot specify reserved emoji ${EMOJI_CONFIG_ERROR}.`);
    });
  });

  describe('duplicate config name', function () {
    beforeEach(function () {
      mockFs({
        'package.json': JSON.stringify({
          name: 'eslint-plugin-test',
          exports: 'index.js',
          type: 'module',
        }),

        'index.js': `
              export default {
                rules: {
                  'no-foo': { meta: { docs: { description: 'Description for no-foo.'} }, create(context) {} },
                },
                configs: {
                  recommended: { rules: { 'test/no-foo': 'error' } },
                }
              };`,

        'README.md': '## Rules\n',

        'docs/rules/no-foo.md': '',

        // Needed for some of the test infrastructure to work.
        node_modules: mockFs.load(PATH_NODE_MODULES),
      });
    });

    afterEach(function () {
      mockFs.restore();
      jest.resetModules();
    });

    it('throws an error', async function () {
      await expect(
        generate('.', {
          configEmoji: [
            ['recommended', 'ðŸ”¥'],
            ['recommended', 'ðŸ˜‹'],
          ],
        })
      ).rejects.toThrow(
        'Duplicate config name in configEmoji options: recommended'
      );
    });
  });

  describe('with one config that does not have emoji', function () {
    beforeEach(function () {
      mockFs({
        'package.json': JSON.stringify({
          name: 'eslint-plugin-test',
          exports: 'index.js',
          type: 'module',
        }),

        'index.js': `
          export default {
            rules: {
              'no-foo': { meta: { docs: { description: 'Description for no-foo.'} }, create(context) {} },
            },
            configs: {
              configWithoutEmoji: { rules: { 'test/no-foo': 'error' } },
            }
          };`,

        'README.md': '## Rules\n',

        'docs/rules/no-foo.md': '',

        // Needed for some of the test infrastructure to work.
        node_modules: mockFs.load(PATH_NODE_MODULES),
      });
    });

    afterEach(function () {
      mockFs.restore();
      jest.resetModules();
    });

    it('shows the default config emoji', async function () {
      await generate('.');
      expect(readFileSync('README.md', 'utf8')).toMatchSnapshot();
      expect(readFileSync('docs/rules/no-foo.md', 'utf8')).toMatchSnapshot();
    });
  });
});
