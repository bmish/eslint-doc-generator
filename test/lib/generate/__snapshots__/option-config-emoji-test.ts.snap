// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generate (--config-emoji) basic shows the correct emojis 1`] = `
"## Rules
<!-- begin auto-generated rules list -->

ðŸ’¼ Configurations enabled in.\\
ðŸ”¥ Set in the \`recommended\` configuration.\\
ðŸŽ¨ Set in the \`stylistic\` configuration.

| Name                           | Description             | ðŸ’¼                               |
| :----------------------------- | :---------------------- | :------------------------------- |
| [no-bar](docs/rules/no-bar.md) | Description for no-bar. | ðŸ”¥ ðŸŽ¨                            |
| [no-baz](docs/rules/no-baz.md) | Description for no-boz. | ðŸŽ¨ ![badge-configWithoutEmoji][] |
| [no-foo](docs/rules/no-foo.md) | Description for no-foo. | ðŸ”¥                               |

<!-- end auto-generated rules list -->
"
`;

exports[`generate (--config-emoji) basic shows the correct emojis 2`] = `
"# Description for no-foo (\`test/no-foo\`)

ðŸ’¼ This rule is enabled in the ðŸ”¥ \`recommended\` config.

<!-- end auto-generated rule header -->
"
`;

exports[`generate (--config-emoji) basic shows the correct emojis 3`] = `
"# Description for no-bar (\`test/no-bar\`)

ðŸ’¼ This rule is enabled in the following configs: ðŸ”¥ \`recommended\`, ðŸŽ¨ \`stylistic\`.

<!-- end auto-generated rule header -->
"
`;

exports[`generate (--config-emoji) basic shows the correct emojis 4`] = `
"# Description for no-boz (\`test/no-baz\`)

ðŸ’¼ This rule is enabled in the following configs: \`configWithoutEmoji\`, ðŸŽ¨ \`stylistic\`.

<!-- end auto-generated rule header -->
"
`;

exports[`generate (--config-emoji) removing default emoji for a config reverts to using a badge for the config 1`] = `
"## Rules
<!-- begin auto-generated rules list -->

ðŸ’¼ Configurations enabled in.

| Name                           | Description             | ðŸ’¼                     |
| :----------------------------- | :---------------------- | :--------------------- |
| [no-foo](docs/rules/no-foo.md) | Description for no-foo. | ![badge-recommended][] |

<!-- end auto-generated rules list -->
"
`;

exports[`generate (--config-emoji) removing default emoji for a config reverts to using a badge for the config 2`] = `
"# Description for no-foo (\`test/no-foo\`)

ðŸ’¼ This rule is enabled in the \`recommended\` config.

<!-- end auto-generated rule header -->
"
`;

exports[`generate (--config-emoji) rule with emoji and badge configs sorts emojis before badges 1`] = `
"## Rules
<!-- begin auto-generated rules list -->

ðŸ’¼ Configurations enabled in.\\
ðŸŽ¨ Set in the \`bar\` configuration.\\
ðŸ”¥ Set in the \`foo\` configuration.

| Name                           | Description             | ðŸ’¼                   |
| :----------------------------- | :---------------------- | :------------------- |
| [no-foo](docs/rules/no-foo.md) | Description for no-foo. | ðŸŽ¨ ðŸ”¥ ![badge-baz][] |

<!-- end auto-generated rules list -->
"
`;

exports[`generate (--config-emoji) rule with emoji and badge configs sorts emojis before badges 2`] = `
"# Description for no-foo (\`test/no-foo\`)

ðŸ’¼ This rule is enabled in the following configs: ðŸŽ¨ \`bar\`, \`baz\`, ðŸ”¥ \`foo\`.

<!-- end auto-generated rule header -->
"
`;

exports[`generate (--config-emoji) with one config that does not have emoji shows the default config emoji 1`] = `
"## Rules
<!-- begin auto-generated rules list -->

ðŸ’¼ Configurations enabled in.

| Name                           | Description             | ðŸ’¼                            |
| :----------------------------- | :---------------------- | :---------------------------- |
| [no-foo](docs/rules/no-foo.md) | Description for no-foo. | ![badge-configWithoutEmoji][] |

<!-- end auto-generated rules list -->
"
`;

exports[`generate (--config-emoji) with one config that does not have emoji shows the default config emoji 2`] = `
"# Description for no-foo (\`test/no-foo\`)

ðŸ’¼ This rule is enabled in the \`configWithoutEmoji\` config.

<!-- end auto-generated rule header -->
"
`;
